### PR TITLE
chisel refactoring to enable split main <-> test

### PR DESCRIPTION
**Pull request by jorendumoulin**
See here my proposals to be able to split testing and source code without errors.
This includes the following:

- remove the emitVerilog statements: they should be implemented by the user of the streamer with the correct parameters (like it is done in #100). If these are useful for testing and we still want to keep them somewhere, I can reimplement them in a test.
- removed most of the default argument assignments, as most of them held an arbitrary value (defaults should only be set if this value will be used in almost all cases)
- some slight argument renaming to enable the split

### CODE REVIEW

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala:1]: Empty file?

File: hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:1]: Also an empty file?

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala:1]: Are you supposed to put this here?

```scala
package snax.csr_manager

import snax.utils._

import chisel3._
import chisel3.util._

// Scala main function for generating CsrManager system verilog file
object CsrManager extends App {
  emitVerilog(
    new CsrManager(
      csrManagerTestParameters.csrNum,
      csrManagerTestParameters.csrAddrWidth,
      csrManagerTestParameters.csrModuleTagName
    ),
    Array("--target-dir", "generated/csr_manager")
  )
}
```

Comment (rgantonio) [reply hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala:1]: When you do, could you try to:

1. Go to the `hw/chisel/` directory
2. Generate the CSR manager with `sbt "runMain snax.csr_manager.CsrManager"` and see if it generates correctly?

File: hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:1]: For this file too make sure to add those I mean one of those and test if they generate properly: ` sbt "runMain snax.streamer.MacStreamerTop "`

```scala
// streamertop for MAC
object MacStreamerTop extends App {
  emitVerilog(
    new StreamerTop(
      StreamerParams(
        temporalAddrGenUnitParams =
          MacStreamerParameters.temporalAddrGenUnitParams,
        fifoReaderParams = MacStreamerParameters.fifoReaderParams,
        fifoWriterParams = MacStreamerParameters.fifoWriterParams,
        stationarity = MacStreamerParameters.stationarity,
        dataReaderParams = MacStreamerParameters.dataReaderParams,
        dataWriterParams = MacStreamerParameters.dataWriterParams
      )
    ),
    Array("--target-dir", "generated/streamertop/mac")
  )
}
```

**Review by rgantonio — COMMENTED** [review]
I feel like it does look cleaner because all test files are collated (collected/aggregated). My only worry is:

1. Will it break when I generate new parameters? I left some comments on parts that i think are critical.
2. Will it also break @xiaoling-yi 's flow? Haha, make sure to discuss together and meet half-way 😄

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala
Comment (jorendumoulin) [reply hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala:1]: aha yes I meant to do this but then I completely forgot, let me see

File: hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+class StreamerTopGenerate
+extends AnyFlatSpec {
+
+  emitVerilog(
+    new StreamerTop(TestParameters.streamer),
+    Array("--target-dir", "generated/")
+  )    
+
+}
```
Comment (xiaoling-yi) [inline hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:15]: I vote to put the parameter example for instancing gemm and simd back. We can write some comments in the code to say this example configuration is for which accelerator. Because I think 1) these example configurations are good for people to understand how to configure the streamer as the configuration format of the streamer is quite complicated. 2) making sure the streamer still works after any modification for these configurations is essential. 3) These configurations show the streamer's reusability, working for different accelerators.

File: hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala
Comment (jorendumoulin) [reply hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:1]: See comment above!

File: hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala (hunk 1 above)
Comment (jorendumoulin) [reply hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:15]: Alright! As discussed: we like to avoid having code that is not used anywhere. Therefore the example configuration is displayed in the added documentation. The tests for different configurations are run through the automatic generation of scala parameters using the .hjson configuration files! For now this is only ALU but when the other accelerators are added these will be tested as well!

File: hw/chisel/doc/streamer.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+### CSR definition
+The offset below is defined by the SNAX core. A more detailed explanation of what are these configurations can be found at `StreamerTop.scala`.
+
+| Address | CSR name             | Notes                               |
+|---------|--------------------------|-------------------------------------|
+| offset + [0..temporalDim - 1]   |    temporalLoopBoundCSRs     | temporal loop bound for each temporal dimension.           |
+| offset + temporalDim + [0..temdataMoverNum * temporalDimporalDim - 1]   |    temporalLoopSrtidesCSRs     | temporal loop strides for each temporal dimension and for each data mover.          |
+| offset + temporalDim +  dataMoverNum * temporalDim + [0..spatialDim.sum - 1]   |    spatialLoopSrtidesCSRs     | spatial loop strides for each data mover and for corresponding spatial dimension. The spatial dimension for each data mover can be different. It depends on the accelerator.        |
+| offset + temporalDim +  dataMoverNum * temporalDim + spatialDim.sum + [0..dataMoverNum - 1]   |    basePtrCSRs     |  base pointers for each data mover.          |
```
Comment (xiaoling-yi) [inline hw/chisel/doc/streamer.md:231]: Add the status CSR here too?

File: hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala (hunk 1 above)
Comment (xiaoling-yi) [inline hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:15]: Can we add the modular system verilog generation test back? It is useful for modular generation and tests.

**Review by xiaoling-yi — COMMENTED** [review]
Overall this PR looks good to me, I just have some small comments. Please kindly check it!

File: hw/chisel/.gitignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
+generated
+project
```
[the remark this answers is not in the document]
Comment (jorendumoulin) [reply hw/chisel/.gitignore]: Aha, yes I missed this, I excluded those two files!

File: hw/chisel/doc/streamer.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+It is written in CHISEL 5.0.0 and is intended to be connected to the SNAX accelerator RISC-V manager core through a SystemVerilog wrapper.
+
+Thanks to the strong expression power of Chisel, the Flexible Streamer Generator has the capability of dealing with any number of temporal loops and any number of parallel loops for data address generation. More importantly, the Streamer generator is accelerator agnostic, highly flexible, and parameterizable which means it is suitable for a wide range of accelerators employing a [tiled-strided-layout](https://github.com/KULeuven-MICAS/snax-mlir/tree/main/compiler/ir/tsl). We have generated Streamers for three accelerators, including the SNAX-GEMM, SNAX-PostProcessing-SIMD, and MAC Engine.
+
+## Microarchitecture
+The microarchitecture of the Flexible Streamer Generator is shown below.
+
+<p align="center">
+  <img src="./docs/microarch.svg" alt="">
+</p>
```
[the remark this answers is not in the document]
Comment (jorendumoulin) [reply hw/chisel/doc/streamer.md:16]: I re-added it!

File: hw/chisel/doc/streamer.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+|  | valid | io_csr_rsp_valid | 1 | Out| The signal indicates if this response is valid. |
+|  | ready | io_csr_rsp_ready | 1 | In| The signal indicates if the SNAX core is ready for this CSR response. |
+| tcdm_req | data | io_data_tcdm_req_0_bits_data | 64| Out| The data from TCDM request. This data is only valuable when it is a write request. |
+| tcdm_req | addr | io_data_tcdm_req_0_bits_addr | 32| Out| The address from TCDM request. |
+| tcdm_req | write | io_data_tcdm_req_0_bits_write | 1| Out| The signal indicates this request is for CSR write or read. |
+| tcdm_req | valid | io_data_tcdm_req_0_valid | 1| Out| The signal indicates if this request is valid. |
+| tcdm_req | ready | io_data_tcdm_req_0_ready | 1| Int| The signal indicates if the TCDM is ready for this CSR request. |
+| | . | . | . | . | There can be a large number of tcdm_req ports depending on the spatial unrolling factors for the data readers and the data writers. tcdm_req ports for readers have lower index number. A detailed mapping for the tcdm_req ports and the data mover ports can be found at `Streamer.scala` line 269-311.|
```
[the remark this answers is not in the document]
Comment (jorendumoulin) [reply hw/chisel/doc/streamer.md]: Yes I just deleted it because it might change later

File: hw/chisel/doc/streamer.md (hunk 2 above)
Comment (jorendumoulin) [reply hw/chisel/doc/streamer.md:231]: Added!

File: hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala (hunk 1 above)
Comment (jorendumoulin) [reply hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:15]: What is the modular system verilog genration?

File: hw/chisel/doc/microarch.svg
Comment (rgantonio) [inline hw/chisel/doc/microarch.svg:1]: Do you think it's better we dump images in our repo? I have images that were dumped in a google drive then sent the link to the gh pages 😅 what do you think?

File: hw/chisel/README.md
Comment (rgantonio) [inline hw/chisel/README.md:1]: I'm actually surprised we put the README's in this PR already. For the CSR Manager we do that in a separate PR?

**Review by rgantonio — COMMENTED** [review]
I will give the go signal here but I prefer Xiaoling gives the final go signal since she uses this more closely than I do 😄 

Otherwise, I am okay with it. Let's just make sure to fill-in the CSR manager docu.

File: hw/chisel/README.md
Comment (jorendumoulin) [reply hw/chisel/README.md:1]: Yes, ideally it should've been included previously, but I never noticed. We need it here because the readme replaces some code that wasn't used and thus removed. For the CSR manager we indeed best separate it. Although is there already documentation for this? The streamer readme is just a copy from the old repo

File: hw/chisel/doc/microarch.svg
Comment (jorendumoulin) [reply hw/chisel/doc/microarch.svg:1]: A lot of people seem to have a lot of opinions about this but in the end it doesn't seem to matter much 😅 . The only worry about Google Drive links is that if they disappear they are gone forever. Maybe the best way is including it as github user content, I'll change it to that!

File: hw/chisel/README.md
Comment (jorendumoulin) [reply hw/chisel/README.md:1]: Thank you!

File: hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala (hunk 1 above)
Comment (jorendumoulin) [reply hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:15]: We do not want people to use subparts as systemverilog components, right?
The testing of systemverilog generation is done by the top module, the only one that we are going to use. All lower modules are included in this test. Having them separate is duplicated testing, and is confusing as we don't want to generate verilog for the modules separately. Does that make sense?

File: hw/chisel/src/test/scala/snax/csr_manager/StreamerTestParameter.scala [hunk 6]
```diff
@@ -90,7 +65,40 @@ object StreamerTestConstant extends CommonParams {
...
+    elementWidth = 8,
+    spatialDim = 2,
+    )
+
+  val fifo = FIFOParams (
+    width = 512,
+    depth = 4
+  ) 
 }
+
```
[the remark this answers is not in the document]
Comment (jorendumoulin) [reply hw/chisel/src/test/scala/snax/csr_manager/StreamerTestParameter.scala]: Yes, this was an error. I fixed it!

File: hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala (hunk 1 above)
Comment (xiaoling-yi) [reply hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:15]: Ok, I understand your point of view. I am fine to merge this PR.

Comment (jorendumoulin) [reply hw/chisel/src/test/scala/snax/streamer/StreamerTopGenerate.scala:15]: Ok, thank you!
